provider/ec2/environ.go

```diff
@@ -269,16 +269,30 @@ func (e *environ) AvailabilityZones(ctx context.ProviderCallContext) (corenetwor
 func (e *environ) InstanceAvailabilityZoneNames(ctx context.ProviderCallContext, ids []instance.Id) ([]string, error) {
 	instances, err := e.Instances(ctx, ids)
 	if err != nil && err != environs.ErrPartialInstances {
-		return nil, err
+		return nil, errors.Trace(err)
 	}
-	zones := make([]string, len(instances))
-	for i, inst := range instances {
+
+	return gatherAvailabilityZones(instances), nil
+}
+
+func gatherAvailabilityZones(instances []instances.Instance) []string {
+	zones := make([]string, 0)
```

Comment:
I think we normally spell this `var zones []string`. See https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices and also:

```
~/w/juju$ rg 'var \w+ \[\]string' | wc -l
439
~/w/juju$ rg ':= make\(\[\]string, 0\)' | wc -l
21
```

Reply:
That said, I think this change of appending only the non-nil ones breaks the contract in the `ErrPartialInstances`, because in the case where `e.Instances` returns partial instances, the length of the returned slice may be less than the input `ids` slice, and the slots in the input slice won't match up to the slots in the returned slice -- seems important for proper usage.

Here are the doc comments for `ZonedEnviron.InstanceAvailabilityZoneNames` and `Environ.Instances` confirming this:

```
	// InstanceAvailabilityZoneNames returns the names of the availability
	// zones for the specified instances. The error returned follows the same
	// rules as Environ.Instances. [see below]

	// Instances returns a slice of instances corresponding to the
	// given instance ids.  If no instances were found, but there
	// was no other error, it will return ErrNoInstances.  If
	// some but not all the instances were found, the returned slice
	// will have some nil slots, and an ErrPartialInstances error
	// will be returned.
```

Note also that the original version returned `err` at the end (the error from `e.Instances`), not `nil`, so as to pass on `ErrPartialInstances` if necessary. Callers that don't handle partial instances will just have the ordinary check for `err != nil`, and will never look at the slice in the case where it has nils / "holes".

We should probably have a unit test for the partial instances case (thanks for adding those tests, BTW).

Reply:
Yeah, I hate it. There are so many things wrong with this. I've made a change that will drop this weird slice and favoring a map with `instance.Id` as the key and the zone names as the values. That way it doesn't matter the ordering or if there are missing slots.

Doing the work highlighted that each provider could potentially do this wrong and lose the nuance of the original intention. As highlighted in my original commit.

That problem should be gone for ever.

Reply:
See #12614